### ISSUE TITLE
Laravel octane random results when using Auth::user()->can('permisssion') after modifying role/permission belonging

### ISSUE THREAD

**Issue by davidjr82** [spatie/laravel-permission#2322]
**Describe the bug**
Using Laravel Octane with roadrunner, if you update a role permission, calling to `Auth::user()->can('permisssion')` leads to unexpected randomly results (randomly grants or denies).

This does not happens with the same code when not running Octane.

Happens both with `Auth::user()->can('permisssion')` and `auth()->user()->can('permisssion')`

**Versions**

- spatie/laravel-permission package version: 5.8.0
- laravel/framework package: 9.48.0
- laravel/octane package: 1.4.0

PHP version: 8.1.15

Database version: mysql 8.0.31

**To Reproduce**

- Install clean laravel app and laravel permissions package
- create a role with a permission, and asign the role to your user
- make a form to modify the belonging of the permission to the role:
```php
        $grant_permission
            ? $role->givePermissionTo($permission)
            : $role->revokePermissionTo($permission);
```
- Put in blade `{{ auth()->user()->can('permission-name') ? 1 : 0 }}` and refresh many times
- You will see the number randomly changing to 0 or 1 

Here is my example code and/or tests showing the problem in my app:

**Expected behavior**
Same code works fine without Octane. With Octane should work in the same way, it should show consistently the "can" of the user.

**Environment (please complete the following information, because it helps us investigate better):**
 - Using sail to develop in WSL2
 - laravel/sail: 1.18.1
 - supervisor line to start octane: `command=/usr/bin/php -d variables_order=EGPCS /var/www/html/artisan octane:start --watch --server=roadrunner --host=0.0.0.0 --rpc-port=6001 --port=80`

**Comment by parallels999** [comment]
Hi, feel free to make a PR to fix that

**Comment by davidjr82** [comment]
I would do, but I have no idea how to fix it 🤣

**Comment by parallels999** [comment]
So, make a test or a minimal laravel aplication showing the problem for somebody review

**Comment by parallels999** [comment]
https://spatie.be/docs/laravel-permission/v5/basic-usage/new-app

**Comment by erikn69** [comment]
Maybe cache is failing, try using array as cache
https://github.com/spatie/laravel-permission/blob/6e0f9574d26d4d7ea16d63508db5b0a44424a862/config/permission.php#L159

**Comment by davidjr82** [comment]
> Maybe cache is failing, try using array as cache
> 
> https://github.com/spatie/laravel-permission/blob/6e0f9574d26d4d7ea16d63508db5b0a44424a862/config/permission.php#L159

Already tested, not related to cache (tested also with array and redis).

**Comment by erikn69** [comment]
I'm not using octane, i can't replicate your problem

**Comment by davidjr82** [comment]
> > This does not happens with the same code when not running Octane.
> 
> I'm not using octane, i can't replicate your problem

I understand, I will try to do a repo adding laravel sail with octane and roadrunner to make it easy to replicate in a sandbox

**Comment by parallels999** [comment]
> Octane boots your application once, keeps it in memory, and then feeds it requests at supersonic speeds.

Maybe it boots different threads, so when you update permissions on one on them, others stay unupdated, and on refresh many times the view changes from the updated one to others, so is there a way to reload octane boots after changing role/permission belonging?

**Comment by davidjr82** [comment]
> > Octane boots your application once, keeps it in memory, and then feeds it requests at supersonic speeds.
> 
> Maybe it boots different threads, so when you update permissions on one on them, others stay unupdated, and on refresh many times the view changes from the updated one to others, so is there a way to reload octane boots after changing role/permission belonging?

It is clearly related to this. I think it has to do with how the gate or auth facade is resolved, like it does only once at the boot time and then never resolves again. Happens the same for example with the container, with Octane you should resolve it with the `app()` helper instead of using dependency injection.

What I don't know is how laravel/permission adds their permission system to the `can` utility. Can be it is not resolving in the "Octane way" and uses some kind of request cache?

**Comment by parallels999** [comment]
> I don't know is how laravel/permission adds their permission system to the can utility. Can be it is not resolving in the "Octane way" and uses some kind of request cache?

There is a PR https://github.com/spatie/laravel-permission/pull/2321 changing that(adds their permission system to the can utility), try it, maybe it fix it

**Comment by olivernybroe** [comment]
A possible issue could be with the `PermissionRegistrar` class. It has multiple static properties being set on it instead of just utilizing the nature of a singleton. 

This could potentially give some issues in the case of octane

**Comment by parallels999** [comment]
> A possible issue could be with the PermissionRegistrar class. It has multiple static properties being set on it instead of just utilizing the nature of a singleton.

It could be a PR for V6

**Comment by drbyte** [comment]
> I think it has to do with how the gate or auth facade is resolved, like it does only once at the boot time and then never resolves again. Happens the same for example with the container, with Octane you should resolve it with the `app()` helper instead of using dependency injection.
> 
> What I don't know is how laravel/permission adds their permission system to the `can` utility. Can be it is not resolving in the "Octane way" and uses some kind of request cache?

It registers it on the gate using the app() helper:

https://github.com/spatie/laravel-permission/blob/6e0f9574d26d4d7ea16d63508db5b0a44424a862/src/PermissionRegistrar.php#L130-L145

**Comment by davidjr82** [comment]
> app(Gate::class)->before(function (Authorizable $user, string $ability) {

@drbyte  Couldn't it be the `Authorizable $user` dependency injection? If it is resolved only once during the boot time and then returns the same cached value, makes sense because all the workers boots once, and only the one who has made the update in the role permission update that change in their memory, so when the app asks for it, if is the worker who made the change gives the actual value, but any other worker will return the original unmodified boot-time value.





> Also try cleaning loaded permissions on some middleware, forcing to read cache again on every call
> 
> https://github.com/spatie/laravel-permission/blob/6e0f9574d26d4d7ea16d63508db5b0a44424a862/src/PermissionRegistrar.php#L162-L165

@olivernybroe Already tried, both flushing whole cache and calling `app()->make(\Spatie\Permission\PermissionRegistrar::class)->forgetCachedPermissions();`. Same result.

**Comment by fredsal** [comment]
> It has multiple static properties being set on it instead of just utilizing the nature of a singleton.

hi @olivernybroe great idea, make a PR please

**Comment by olivernybroe** [comment]
> > It has multiple static properties being set on it instead of just utilizing the nature of a singleton.
> 
> hi @olivernybroe great idea, make a PR please

https://github.com/spatie/laravel-permission/pull/2324

**Comment by drbyte** [comment]
`dev-master` (v`6.x`) now has the refactor by @olivernybroe 

Feedback welcome!

**Note the need to replace your migration files as part of the change.**

/ping @davidjr82

**Comment by olivernybroe** [comment]
@drbyte there might be more changes needed for full octane support. But that change should at least help a bit.

**Comment by drbyte** [comment]
> @drbyte there might be more changes needed for full octane support. But that change should at least help a bit.

Thanks. Appreciate your support.

Before Octane came along we made a few refinements for long-running instances, but since I'm not using that in my tech stack right now it's not one of my strengths these days.

Will keep this open as we get feedback from testing, and collaborate to ensure compatibility.

**Comment by drbyte** [comment]
Thanks @olivernybroe. Appreciate your support.

**Comment by davidjr82** [comment]
@drbyte tested dev-main branch publishing migration and config file, and clearing all caches, and same result.

@olivernybroe thanks for the PR. Even if doesn't resolves the issue, it is a needed step to have a good Octane integration.

I hope to have some time this weekend to create a repo with the issue, using sail and roadrunner. What I don't know at the moment is how to create a test specifically for Octane (this same issue in the PHP builtin server works).

**Comment by davidjr82** [comment]
I am not able to reproduce the issue in a simple repo, therefore there is something in my code that affects the behaviour that I am missing. 

Although it is something related to Octane -because when running in the builtin php server the issue disappears-, I propose to close this issue and let someone that is able to reproduce an Octane issue to open a new one.

If you want to keep this issue open to address the Octane compatibility for 6.x, good for me. But we would need someone that can recreate the issue in a separate and simple repo to address the right problem.

In my case, I am going to stop using Octane for the moment and stick to a regular PHP server. 
Thanks to you all for your help.

**Comment by Cluster2a** [comment]
We also plan to use this package on an octane-driven Laravel 9 instance.
Are we good to go, or are there still potential issues with this package?

Should I use the dev-master for octane?

**Comment by davidjr82** [comment]
I personally have dropped Octane in my project to avoid conflicts with this package. I still have the issues with the dev-master branch.

**Comment by erikn69** [comment]
It would not be possible to add some observer that in the `change permissions` event reloads the cache/logged user data of the entire service

also, did you try adding a middleware to force permissions reload on every call?
```php
<?php

namespace App\Http\Middleware;

use Closure;

class ChangeDatabase{
    /**
     * Handle an incoming request.
     *
     * @param  \Illuminate\Http\Request  $request
     * @param  \Closure  $next
     * @return mixed
     */
    public function handle($request, Closure $next){
        $user=auth('api')->user();
        if (is_null($user)) 
            $user=auth()->user();

        if (!empty($user)) {
            $user->load('roles', 'permissions');
        }

        return $next($request);
    }
}

```
please, try that and tell us, thanks

**Comment by erikn69** [comment]
Maybe a better solution here: https://github.com/mcamara/laravel-localization/issues/780#issuecomment-981962203, https://github.com/mcamara/laravel-localization/issues/780#issuecomment-1072084423

1. Add a listener to hook into Octane's RequestReceived event
```php
namespace App\Listeners;

use Laravel\Octane\Events\RequestReceived;
use Spatie\Permission\PermissionRegistrar;

class ReloadPermissions
{
    public function handle(RequestReceived $event): void
    {
        app()->make(PermissionRegistrar::class)->clearClassPermissions();
    }
}
```
2. Reference the listener in Octane's config file:
```php
'listeners' => [
        RequestReceived::class => [
            ...Octane::prepareApplicationForNextOperation(),
            ...Octane::prepareApplicationForNextRequest(),
            \App\Listeners\ReloadPermissions::class
        ],
```

**Comment by parallels999** [comment]
> Is there anything else I can try to reproduce this issue?

@Cluster2a the supposed problem is when revoking/changing permissions on a role, and making many calls, so:

- create a role R1
- create permissions
- assign permissions to role R1
- assign Role to user
- make one request to get that user permissions
- remove all permissions from R1 `$role->syncPermissions([]);`
- make many requests to get `{{ auth()->user()->can('your-permission-name') ? 1 : 0 }}`

**Comment by parallels999** [comment]
Maybe it is just @davidjr82 implementation problem

**Comment by davidjr82** [comment]
@Cluster2a maybe is my implementation. Have you tried to put one request waiting after the change (like sleep(20)) and check within other request?

@parallels999 maybe it is my implementation, but when I change sail to use the default php built-in server instead of roadrunner, the problem disappears.

But I am fine closing this issue because as I said I have dropped Octane to use this package.

I think it makes more sense to open a new issue when someone can replicate this kind of problems in a small environment, so we can solve it.

**Comment by parallels999** [comment]
> I think it makes more sense to open a new issue when someone can replicate this kind of problems in a small environment

I think that would be the most appropriate action

**Comment by davidjr82** [comment]
@Cluster2a I finally discovered that the issue was produced by spatie/once package. Just to make you know so you don't drop this package to use Octane because of this.

**Comment by drbyte** [comment]
@davidjr82 Thanks for updating this!

**Comment by parallels999** [comment]
There is a PR now https://github.com/spatie/laravel-permission/pull/2403
